### PR TITLE
MON-3700: replace OAuth proxy for Thanos Ruler

### PR DESCRIPTION
**Pull request by simonpasquier**
* [ ] I added CHANGELOG entry for this change.
* [X] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by Tai-RedHat** [comment]
PR tested with cluster-bot
/label qe-approved

File: assets/thanos-ruler/service-account.yaml [hunk 1]
```diff
@@ -1,8 +1,7 @@
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  annotations:
-    serviceaccounts.openshift.io/oauth-redirectreference.thanos-ruler: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"thanos-ruler"}}'
+  annotations: {}
```
Comment (machine424) [inline assets/thanos-ruler/service-account.yaml:4]: I'm afraid we'll need to do it in two steps https://github.com/openshift/cluster-monitoring-operator/blob/e3ea72fdc49d14443583418002e25b136f69d692/Documentation/development.md?plain=1#L52
(just realized that section is in the wrong place, I'll move, tell me if you have any preference.)
(I didn't test/check this, ignore if you did)

File: pkg/tasks/thanos_ruler_user_workload.go [hunk 2]
```diff
@@ -441,6 +432,11 @@ func (t *ThanosRulerUserWorkloadTask) destroy(ctx context.Context) error {
 		return fmt.Errorf("deleting ThanosRuler object failed: %w", err)
 	}
 
+	err = t.client.DeleteSecret(ctx, acs)
+	if err != nil {
+		return fmt.Errorf("creating Thanos Ruler alertmanager config Secret failed: %w", err)
```
Comment (machine424) [inline pkg/tasks/thanos_ruler_user_workload.go:437]:
```suggestion
		return fmt.Errorf("deleting Thanos Ruler alertmanager config Secret failed: %w", err)
```

File: pkg/tasks/thanos_ruler_user_workload.go [hunk 3]
```diff
@@ -291,6 +278,35 @@ func (t *ThanosRulerUserWorkloadTask) create(ctx context.Context) error {
 		}
 	}
 
+	// TODO(simonpasquier): remove this step after OCP 4.16 is released.
```
Comment (machine424) [inline pkg/tasks/thanos_ruler_user_workload.go:281]: If it's not already the case, it'll be great to have a ticket to keep track of thee TODOs.

File: assets/thanos-ruler/thanos-ruler.yaml [hunk 4]
```diff
@@ -76,8 +60,10 @@ spec:
     volumeMounts:
     - mountPath: /etc/tls/private
       name: secret-thanos-ruler-tls
-    - mountPath: /etc/proxy/secrets
-      name: secret-thanos-ruler-oauth-cookie
+      readOnly: true
+    - mountPath: /etc/kube-rbac-proxy
```
Comment (machine424) [inline assets/thanos-ruler/thanos-ruler.yaml:64]: I don't know if there is a mechanism to "reload" the proxies if such secrets change without the deployments being changed. (or if we need to pair them, it's not as critical as the certs as we'll be making the changes thus we'll test them, but we may overlook things)

File: pkg/tasks/thanos_ruler_user_workload.go [hunk 5]
```diff
@@ -174,22 +177,6 @@ func (t *ThanosRulerUserWorkloadTask) create(ctx context.Context) error {
 	}
 
 	{
-		// Create trusted CA bundle ConfigMap.
-		trustedCA, err := t.factory.ThanosRulerTrustedCABundle()
```
Comment (machine424) [inline pkg/tasks/thanos_ruler_user_workload.go:178]: I think we can get rid of `ThanosRulerTrustedCABundle()` and its deps.

**Review by machine424 — CHANGES_REQUESTED** [review]
Thanks for the cleaning

File: assets/thanos-ruler/thanos-ruler.yaml (hunk 4 above)
Comment (machine424) [reply assets/thanos-ruler/thanos-ruler.yaml:64]: (not related to this PR, we can discuss/address this later)

Comment (simonpasquier) [reply assets/thanos-ruler/thanos-ruler.yaml:64]: kube-rbac-proxy should live-reload certificates

File: pkg/tasks/thanos_ruler_user_workload.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/tasks/thanos_ruler_user_workload.go:281]: I'll create one

File: assets/thanos-ruler/thanos-ruler.yaml (hunk 4 above)
Comment (machine424) [reply assets/thanos-ruler/thanos-ruler.yaml:64]: yep, I was talking about the proxy config itself.
Whether changing `thanos-user-workload-kube-rbac-proxy-web` in the future requires a manual reloading of the proxy or if we should add a hash or sth as we do for some certs for others components.

File: assets/thanos-ruler/service-account.yaml (hunk 1 above)
Comment (simonpasquier) [reply assets/thanos-ruler/service-account.yaml]: thanks, updated!

Comment (machine424) [reply assets/thanos-ruler/service-account.yaml]: I think `-` should be a suffix :)
I'll add an example as well. I think`kubectl label` uses the same syntax.

File: pkg/tasks/thanos_ruler_user_workload.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/tasks/thanos_ruler_user_workload.go:281]: https://issues.redhat.com/browse/MON-3801

File: assets/thanos-ruler/thanos-ruler.yaml (hunk 4 above)
Comment (machine424) [reply assets/thanos-ruler/thanos-ruler.yaml:64]: NP, I'll create a ticket for that.

**Comment by machine424** [comment]
Thanks.
/lgtm

**Pull request merged**